htdocs/core/lib/files.lib.php

```diff
@@ -1122,6 +1122,117 @@ function dol_move($srcfile, $destfile, $newmask = '0', $overwriteifexists = 1, $
 	return $result;
 }
 
+
+/**
+ * Rename a file or a name, with overwrite.
+ *
+ * The PHP rename method works differently between Linux and Windows platforms.
+ * This method aims to match the linux implementation
+ *
+ * @param	string  $oldname            Source file or directory
+ * @param   string	$newname            Destination file or directory
+ * @return  bool    true if success
+ */
+function dol_rename($oldname, $newname)
+{
+	if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
+		// If Windows, use custom function to rename directories
+		if (is_dir($oldname)) {
+			return dol_win_dir_rename($oldname, $newname);
+		} else {
+			return (bool) @rename($oldname, $newname);
+		}
+	} else {
+		// For Linux and other Unix-like systems, use rename function directly
+		return (bool) @rename($oldname, $newname);
+	}
+}
+
+
+/**
+ * Rename a directory on windows platform
+ *
+ * @param	string  $oldname            Source directory
+ * @param   string	$newname            Destination directory
+ * @return  bool    true if success
+ */
+function dol_win_dir_rename($oldname, $newname)
```

Comment:
The rename on windows should work like on linux. If you need a copy and delete, it probably means you make the move on a different file system ?

Reply:
The PHPUNIT test tries to do it on the same file system (in the documents directory) and that fails.

You can check on [php.net's rename documentation page](https://www.php.net/manual/en/function.rename.php) that the behavior is not the same for all systems.

Reply:
copy and delete is a different operation than rename. If file is large, difference is very important.
I do not experience trouble on windows with rename, both on FAT or NTFS. So we should have no reason to have a special rename function for windows.
May be we can add a test inside dol_move_dir(() to be sure to return the same result. My understanding is that it works the same, but return code differs ?

Reply:
The function '@rename' fails on a windows in a testcase where the destination directory is empty - the result on linux is different.  The function can be updated to move the files rather than copy/delete them.

Reply:
It means you can't move a file called /aaa/filename  into a directory /bbb if /bbb is empty on windows ?